Add DepolarizedAnalyserTransmission Algorithm

cailafinn opened the pull request:
Part of the Polarised SANS Epic. 

### Description of work

#### Summary of work
Adds an algorithm which determines the transmission of a depolarised He3 cell from a depolarised run and an empty cell run. 

Fixes #36383 

#### Further detail of work
Normalises the depolarised workspace by the empty cell workspace and then performs an exponential fit to determine the T_E (empty cell transmission) and pxd (gas depth * pressure). This can then be used in later algorithms to determine the polarisation efficiency for a given wavelength. 

Both input workspaces need to contain a single spectrum and be in wavelength. 

### To test:
1. This basic example from the documentation will provide a decent example of the algorithm in action:
```
CreateSampleWorkspace(OutputWorkspace='mt', Function='User Defined', UserDefinedFunction='name=UserFunction, Formula=1.465e-07*exp(0.0733*4.76*x)', XUnit='wavelength', NumBanks=1, BankPixelWidth=1, XMin=3.5, XMax=16.5, BinWidth=0.1)
CreateSampleWorkspace(OutputWorkspace='dep', Function='User Defined', UserDefinedFunction='name=UserFunction, Formula=0.0121*exp(-0.0733*10.226*x)', XUnit='wavelength', NumBanks=1, BankPixelWidth=1, XMin=3.5, XMax=16.5, BinWidth=0.1)

output = SANSCalcDepolarisedAnalyserTransmission("dep", "mt")

print("PXD Value = " + str(output.column("Value")[0]) + ".")
print("T_E Value = " + str(output.column("Value")[1]) + ".")
```
Note: If you want to test with real data to test with, message me for some. 
2. Check the documentation is formatted properly and can be accessed from the `?` in the algorithm dialog. (You may need to build the `docs-qthelp` target for this to work. 
3. Try the algorithm with some *terrible* fit starting params. It should fail with a runtime error. (Note: The algorithm is designed to fail fast and throw if something goes wrong rather than warning. This is often done near the beginning of a polarisation correction, so it being wrong would throw off a lot).
4. Try the algorithm with an all 0 workspace as the `mt` file. This should also fail with a helpful error message. 

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

Commit:
Add initial blank files

RE #36383

Commit:
Add initial failing tests for boilerplate methods

RE #36383

Commit:
Add a summary

RE #36383

Commit:
Add the constants required for the fit

RE #36383

Commit:
Add a catagory

RE #36383

Commit:
Add initial algorithm properties

RE #36383

Commit:
Divide the depolarised ws by the empty cell ws

RE #36383

Commit:
Add functionality for fitting divided workspace

RE #36383

Commit:
Fix formatting

RE #36383

Commit:
Replace use of ws names with pointers

RE #36383

Commit:
Add starting fit values

RE #36383

Commit:
Fix incorrect sign on conversion factor.

RE #36383

Commit:
Fix override warning

RE #36383

Commit:
Fix issues when providing a non-monitor workspace

Bascically adding a check for a terrible fit caused when the whole
workspace is NaNs or infs. This seems to happen when the MT workspace is
a non-monitor one and therefore contains 0s. This causes a div by zero
error to be created in the output and therefore a NaN value. This can't
be fitted, so the Fit algorithm (which needs to ignore NaNs and Infs to
work properly) just creates a curve using the starting values but has
nothing to fit to. So we add a check for a fit quality of 0 and throw an
error as if the fit failed.

RE #36383

Commit:
Add normal execution test

RE #36383

Commit:
Add failing state test cases

RE #36383

Commit:
Generalise naming convention

RE #36383

Commit:
Add documentation

Also updates some of the in-algorithm documentation to make it a bit
clearer.

RE #36383

Commit:
Add a release note

RE #36383

Commit:
Fix alg name in doctest

RE #36383

Commit:
Make property match coding standards

Shouldn't contain underscores.

RE #36383

Commit:
Add validation for number of spectra and unit

RE #36383

Commit:
Add matching  bin width validation

RE #36383

Commit:
Output calculated fit workspace

RE #36383

Commit:
Update documentation tests & details

RE #36383

Commit:
Make provision of fit curves output workspace optional

RE #36383

Commit:
Calculate the non-normalised covariance matrix

RE #36383

Commit:
Update documentation

RE #36383

Commit:
Sort alg properties into groups

RE #36383

Commit:
Use proportional deltas in tests

RE #36383

Commit:
Add a fit quality override param

RE #36383

rbauststfc reviewed (commented):
I've not had a chance to test the functionality yet, but the code looks good - it's structured nicely and on the whole it's quite easy to follow what the algorithm is doing.

I just had a few questions/comments at this stage. Will follow-up when I've had a chance to test (sorry for the two-part review!).

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/DepolarizedAnalyserTransmission.cpp:
```diff
@@ -0,0 +1,215 @@
+                  "The name of the workspace containing the calculated fit curve.");
+  declareProperty(std::make_unique<WorkspaceProperty<ITableWorkspace>>(
+                      std::string(PropNames::OUTPUT_COV_MATRIX), "", Kernel::Direction::Output, PropertyMode::Optional),
+                  "The name of the table workspace containing the normalised covariance matrix from the fit.");
```
Sorry for getting confused, but is this outputting the normalised covariance matrix, or the non-normalised matrix?

cailafinn replied:
This is the non-normalised one. So good spot on the doc being wrong!

rbauststfc commented:
Perhaps we could walk through a test with the data Diego provided at some point? I've just been having a look at the data and it looks to me like the depolarised runs are workspace groups but the empty cell runs are single workspaces? I tried running the algorithm with the PR test example and it completed OK. I then tried running it by passing a workspace group to the `DepolarisedWorkspace` property and a single workspace for the `EmptyCellWorkspace` and I'm afraid Mantid crashed.

rbauststfc commented:
Yes, I just noticed that as well. It's no problem if that bit should be outside the algorithm, but the crashing when workspace groups are passed in probably still needs a look either way.

Commit:
Create entire fitting function in method

RE #36383

Commit:
Replace s with z in polarisation

We just want to keep consistency with the other algorithms in mantid if
we can.

RE #36383

cailafinn commented on Framework/Algorithms/src/PolarizationCorrections/DepolarizedAnalyserTransmission.cpp:
```diff
@@ -0,0 +1,215 @@
+  fitAlg->setProperty("Function", func);
+  fitAlg->setProperty("InputWorkspace", inputWs);
+  fitAlg->setProperty("IgnoreInvalidData", true);
+  fitAlg->setProperty("StartX", FitValues::START_X);
```
This is taken from Diego's scripts and since it's in wavelength it should be a fairly solid known where the fit should take place. I'll check to make sure they don't want a parameter for this though.

Commit:
Validate monitor workspaces

Ensure that we're checking monitors, as that's the only expected input.

RE #36383

Commit:
Clarify non-normalised TableWorkspace output process

RE #36383

Commit:
Refactor tests

RE #36383

Commit:
Fix crash when given a workspace group

RE #36383

Commit:
Add properties for start and end X

RE #36383

Commit:
Remove monitor checks.

This doesn't work for LARMOR, which uses a summed & masked section of
the main detector to take the place of the post-sample monitor it
doesn't have. Therefore it's not going to have a monitor spectrum.

RE #36383

rbauststfc reviewed (commented):
When running the example script, if I change the algorithm call to:

```
output = DepolarizedAnalyserTransmission("dep", "mt", OutputWorkspace="out", OutputCovarianceMatrix="out_cov", OutputFitCurves="out_fit")
```

I get an error from the lines further down because the variable `output` has type `mantid.simpleapi.DepolarizedAnalyserTransmission_returns`. I've tried checking the behaviour with algorithm `CreateLogPropertyTable`, which also has a `TableWorkspace` type "OutputWorkspace" property, and it doesn't seem to do the same thing. I've not looked further into why this might be happening I'm afraid, but I think it would be good to see if we can get to the bottom of it.

I've done a thorough test otherwise, checking passing in all sorts of different parameter values, and the algorithm completes successfully when possible and provides appropriate error messages when there's an issue. The `OverrideFitQualityError` option works well too. Other than a couple of very minor comments, the code is looking good too.

Commit:
Remove reference to monitor workspace from  error message

RE #36383

cailafinn commented:
> I get an error from the lines further down because the variable `output` has type `mantid.simpleapi.DepolarizedAnalyserTransmission_returns`. I've tried checking the behaviour with algorithm `CreateLogPropertyTable`, which also has a `TableWorkspace` type "OutputWorkspace" property, and it doesn't seem to do the same thing. I've not looked further into why this might be happening I'm afraid, but I think it would be good to see if we can get to the bottom of it.

I've found this in `LoadSpiceAscii` as well, so seems to be a more general mantid issue/design choice.

rbauststfc commented:
OK, thanks, it looks like `LoadSpiceAscii` has multiple output properties too. I guess if you have multiple output properties then you don't know which should be assigned to the variable, so perhaps that's why. @jclarkeSTFC, FYI, we don't need to worry about the return type errors on these algorithms.

Commit:
Improve quality override naming

RE #36383

rbauststfc reviewed (approved):
All review comments have now been addressed and the code structure is looking good and working well. It will be great to get this into the Nightly so that our scientists can start doing some testing with it - thanks very much for getting this all sorted  👍

rbauststfc commented:
Just as a reminder, I think we should open an issue to add system tests for these new algorithms, but I suggest we write those as a maintenance task after our scientists have had a chance to do some thorough testing and confirm the calculations, etc.
